hardware_interface/src/resource_manager.cpp

```diff
@@ -756,50 +793,83 @@ ResourceManager::ResourceManager(
+        "in the URDF.",
+        individual_hardware_info.name.c_str());
+      components_are_loaded_and_initialized_ = false;
+      break;
+    }
+
+    if (individual_hardware_info.type == actuator_type)
+    {
+      std::scoped_lock guard(resource_interfaces_lock_, claimed_command_interfaces_lock_);
+      if (!resource_storage_->load_and_initialize_actuator(individual_hardware_info))
       {
-        std::scoped_lock guard(resource_interfaces_lock_, claimed_command_interfaces_lock_);
-        resource_storage_->load_and_initialize_actuator(individual_hardware_info);
+        components_are_loaded_and_initialized_ = false;
+        break;
       }
-      if (individual_hardware_info.type == sensor_type)
+    }
+    if (individual_hardware_info.type == sensor_type)
+    {
+      std::lock_guard<std::recursive_mutex> guard(resource_interfaces_lock_);
+      if (!resource_storage_->load_and_initialize_sensor(individual_hardware_info))
       {
-        std::lock_guard<std::recursive_mutex> guard(resource_interfaces_lock_);
-        resource_storage_->load_and_initialize_sensor(individual_hardware_info);
+        components_are_loaded_and_initialized_ = false;
+        break;
       }
-      if (individual_hardware_info.type == system_type)
+    }
+    if (individual_hardware_info.type == system_type)
+    {
+      std::scoped_lock guard(resource_interfaces_lock_, claimed_command_interfaces_lock_);
+      if (!resource_storage_->load_and_initialize_system(individual_hardware_info))
       {
-        std::scoped_lock guard(resource_interfaces_lock_, claimed_command_interfaces_lock_);
-        resource_storage_->load_and_initialize_system(individual_hardware_info);
+        components_are_loaded_and_initialized_ = false;
+        break;
       }
     }
   }
 
-  // throw on missing state and command interfaces, not specified keys are being ignored
-  if (validate_interfaces)
+  if (components_are_loaded_and_initialized_ && validate_storage(hardware_info))
+  {
+    std::lock_guard<std::recursive_mutex> guard(resources_lock_);
+    read_write_status.failed_hardware_names.reserve(
+      resource_storage_->actuators_.size() + resource_storage_->sensors_.size() +
+      resource_storage_->systems_.size());
+  }
+  else
   {
-    validate_storage(hardware_info);
+    std::scoped_lock guard(resource_interfaces_lock_, claimed_command_interfaces_lock_);
+    resource_storage_->clear();
+    // cleanup everything
+    components_are_loaded_and_initialized_ = false;
   }
 
-  std::lock_guard<std::recursive_mutex> guard(resources_lock_);
-  read_write_status.failed_hardware_names.reserve(
-    resource_storage_->actuators_.size() + resource_storage_->sensors_.size() +
-    resource_storage_->systems_.size());
+  return components_are_loaded_and_initialized_;
 }
 
-bool ResourceManager::is_urdf_already_loaded() const { return is_urdf_loaded__; }
+bool ResourceManager::are_components_initialized() const
+{
+  return components_are_loaded_and_initialized_;
```

Comment:
```suggestion
  return (
    !resource_storage_->actuators_.empty() || !resource_storage_->sensors_.empty() ||
    !resource_storage_->systems_.empty());
```

Does it make sense to check the size of the components instead?

Reply:
Maybe, but I would optimize this in a follow-up. I am not sure of all the consequences of this method.

Reply:
ok

Reply:
If we do this this way, we may avoid the gazebo's PR. I'm not sure how we can set this `components_are_loaded_and_initialized_` variable to `true` when the `load_and_initialize_components` cannot be from gazebo as it sets up its own components